### ISSUE TITLE
Decorate maven dependencies in generated pom file

### ISSUE THREAD

**Issue by juergen-albert** [bndtools/bnd#3854]
BND has a nice mechanism to write build dependencies in the pom files it generates in a Bundle. We unfortunately have some dependencies hat are not originally in any maven repository, but have pom.xmls included that name e.g. a SNAPSHOT version for actual Release Versions.   We try to get this fixed in the projects themselves, but this can be hard and takes time. To circumvent this issue I'd like to implement a decoration for such dependencies.

Idea 1: 
We already have the ```-maven-dependencies``` instruction to name additional dependencies, that are can merge or replace the build dependencies. We can use this to overwrite specific build dependencies. Right now we use the absolute file path as namespace for the maven dependencies generated from the build dependencies.

Idea 2:
Create a new instruction, that does this.

Idea 3: 
If the  Artifact came from any maven repository we store the GAV it was retrieved with in the index file. Maybe we can create an instruction to use this information instead of the pom.properties in the jar.

**Comment by bjhargrave** [comment]
If idea 1 already works, just use that. Or some tweaks if necessary. (Can you be more specific on your problem here as I am not super clear on what your issue is?)

Idea 2 seems not a good choice since idea 1 is already there.

I don't see how idea 3 can work since the GAV in the index file retrieves an artifact which should have the identical GAV in its pom. And by the time the builder is building a bundle, it only has a classpath of local files and that is all disconnected from what ever repo sourced the files.

**Comment by juergen-albert** [comment]
We use a couple of bundles that are not available as Maven Artifacts. In our concrete case these are Bundles from Eclipse Updatesites. As most of them are build with maven tycho, they pack a pom in the bundle but they still have the SNAPSHOT Qualifier in there. We bulkupload what we need to our nexus. When we manually upload them, Nexus does not check if the internal pom is in accordance with the given GAV coordinates. 

> If idea 1 already works, just use that. Or some tweaks if necessary. (Can you be more specific on your problem here as I am not super clear on what your issue is?)

As a concrete Example I'd like to to the following

```
-buildpath:\
   osgi.core;version=7.0,\
   org.eclipse.emf.common;version=2.16

-maven-depedencies.emf:\
  org.eclipse.emf.common;\
    groupId=org.eclipse.emf;\
    artifactId=org.eclipse.emf.common;\
    version=2.16.0
```
I wont do avoid to declare every build dependency as maven dependency as well. The example would right now result in 3 dependencies in maven. One for osgi.core, org.eclipse.emf.common (with the SNAPSHOT because of the false pom) and org.eclipse.emf.common from my manual set dependency

Internally we translate the buildpath dependency in something like this:
/user/home/.m2/repositories/org/eclipse/emf/org.eclipse.emf.common.jar;\
    groupId=org.eclipse.emf;\
    artifactId=org.eclipse.emf.common;\
    version=2.16.0

Thus it would be hard to overwrite this one specifically because I would need to know the absolut path to the dependency (or might this be possible with a macro?)

> 
> Idea 2 seems not a good choice since idea 1 is already there.
> 
> I don't see how idea 3 can work since the GAV in the index file retrieves an artifact which should have the identical GAV in its pom. And by the time the builder is building a bundle, it only has a classpath of local files and that is all disconnected from what ever repo sourced the files.
See my explanation from the current Situation. I know that this is kind of a special case, but with the Eclipse Stuff and the shitty builds with tycho, the case is not that rare.

**Comment by bjhargrave** [comment]
I think the PR will address your needs. With the PR, you can fix the example dependency with:

```properties
-maven-depedencies.emf:\
  org.eclipse.emf:org.eclipse.emf.common:2.16.0-SNAPSHOT;\
    groupId=org.eclipse.emf;\
    artifactId=org.eclipse.emf.common;\
    version=2.16.0
```

Note that another solution is to specify the `-maven-dependencies` instruction instead of a merged variant. If the project already specifies a `-maven-dependencies` instruction, the ProjectBuilder will not compute it. But then you don't get the benefit of the remainder of the dependencies being computed.

**Comment by juergen-albert** [comment]
Perfect. Thx BJ!

**Comment by bjhargrave** [comment]
@juergen-albert With the fix in #4020, you can now fix your dependency's maven metadata right in the `-buildpath`!

```properties
-buildpath:\
   osgi.core;version=7.0,\
   org.eclipse.emf.common;version=2.16;\
      maven-groupId=org.eclipse.emf;\
      maven-artifactId=org.eclipse.emf.common;\
      maven-version=2.16.0
```

Note: You must provide all 3 `maven-` attributes. No skimping!

So you will no longer need the `-maven-dependencies.emf` statement example above.
